### PR TITLE
Properly handle MapReduce operation with non-empty total input and empty input after sampling

### PR DESCRIPTION
**Pull request by gritukan**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: yt/yt/server/controller_agent/controllers/sort_controller.cpp [hunk 1]
```diff
@@ -2576,7 +2577,8 @@ class TSortControllerBase
     {
         if (SimpleSort) {
             UpdateTask(SimpleSortTask);
-        } else {
+        }
+        if (IntermediateSortTask) {
```
Comment (psushin) [inline yt/yt/server/controller_agent/controllers/sort_controller.cpp:2581]: It is rather suspicioius, that `UpdateTask(FinalSortTask)`  goes inside this if..

Comment (gritukan) [reply yt/yt/server/controller_agent/controllers/sort_controller.cpp:2581]: Fixed. Btw, there is a common pattern
```
if (SomeTask) {
    UpdateTask(SomeTask);
}
```
in controllers. What do you think about refactoring (in a separate PR) allowing UpdateTask to handle nullptr?

Comment (psushin) [reply yt/yt/server/controller_agent/controllers/sort_controller.cpp:2581]: Soungs good!)

**Comment by psushin** [comment]
Shipped

**Comment by psushin** [comment]
Meh, there are some compat tests failing


`yt/yt/tests/integration/compat/trunk_vs_23_2
test_map_reduce_operation.py::TestMapReduceCommandsCompatNewNodes[7]
test_empty_input_due_to_sampling (gritukan)
`

`
YT_VERIFY(denominator != 0) at yt/yt/core/misc/numeric_helpers-inl.h:18 in DivCeil
*** Aborted at 1701955311 (Unix time); Try "date -d @1701955311" if you are using GNU date ***
*** Begin codicils ***
OperationId: b1a61119-a91293f9-a03e8-da1f6cd
AuthenticatedUser: root
*** End codicils ***
0x00000cf9e97c in NYT::NDetail::AssertTrapImpl(TBasicStringBuf<char, std::__y1::char_traits<char> >, TBasicStringBuf<char, std::__y1::char_traits<char> >, TBasicStringBuf<char, std::__y1::char_traits<char> >, int, TBasicStringBuf<char, std::__y1::char_traits<char> >) at /home/teamcity/source/Yt_ArcRelease/yt/yt/core/misc/assert.cpp:58
*** SIGILL (Illegal instruction) (@0xcf9e97d) received by PID 192 (FID 0xfffef2de5baf53ef TID 0x7fdedcaa5700) from PID 217704829 code ILL_ILLOPN (Illegal operand.) ***

ERR no page found read kernel-mode access
TRAPNO X86_TRAP_UD (Invalid Opcode)
R8 0x567f22c6
R9 0x7fdedf727f80
R10 0x7fdedf727f80
R11 0x287
R12 0x15127987a7d8
R13 0x15127987a700
R14 0x7fdedf728008
R15 0x2
RDI 0x7ffccedf9080
RSI 0xfadcbaaf08d
RBP 0x7fdedf728420
RBX 0x53
RDX 0x35c0f
RAX 0x15127fc10000
RCX 0x7fdf03155927
RSP 0x7fdedf727ff0
RIP 0xcf9e97d
EFL 0x10206
CSGSFS 0x2b000000000033
OLDMASK 0x0
CR2 0x0
 1. 0x00000cf9e97c in NYT::NDetail::AssertTrapImpl(TBasicStringBuf<char, std::__y1::char_traits<char> >, TBasicStringBuf<char, std::__y1::char_traits<char> >, TBasicStringBuf<char, std::__y1::char_traits<char> >, int, TBasicStringBuf<char, std::__y1::char_traits<char> >) at /home/teamcity/source/Yt_ArcRelease/yt/yt/core/misc/assert.cpp:58
 2. 0x0000156b1187 in long NYT::DivCeil<long>(long const&, long const&) at /home/teamcity/source/Yt_ArcRelease/yt/yt/core/misc/numeric_helpers-inl.h:18
 3. 0x0000156b1187 in NYT::NControllerAgent::TPartitionJobSizeConstraints::TPartitionJobSizeConstraints(NYT::TIntrusivePtr<NYT::NScheduler::TSortOperationSpecBase> const&, NYT::TIntrusivePtr<NYT::NControllerAgent::TSortOperationOptionsBase> const&, NYT::NLogging::TLogger, long, long, long, double) at /home/teamcity/source/Yt_ArcRelease/yt/yt/server/controller_agent/job_size_constraints.cpp:680
 4. 0x0000156b0c32 in NYT::TRefCountedWrapper<NYT::NControllerAgent::TPartitionJobSizeConstraints>::TRefCountedWrapper<NYT::TIntrusivePtr<NYT::NScheduler::TSortOperationSpecBase> const&, NYT::TIntrusivePtr<NYT::NControllerAgent::TSortOperationOptionsBase> const&, NYT::NLogging::TLogger&, long&, long&, long&, double&>(NYT::TIntrusivePtr<NYT::NScheduler::TSortOperationSpecBase> const&, NYT::TIntrusivePtr<NYT::NControllerAgent::TSortOperationOptionsBase> const&, NYT::NLogging::TLogger&, long&, long&, long&, double&) at /home/teamcity/source/Yt_ArcRelease/library/cpp/yt/memory/new-inl.h:43
 5. 0x0000156a997a in NYT::TRefCountedWrapper<NYT::NControllerAgent::TPartitionJobSizeConstraints>* NYT::NDetail::TConstructHelper<NYT::NControllerAgent::TPartitionJobSizeConstraints, true>::Construct<NYT::TIntrusivePtr<NYT::NScheduler::TSortOperationSpecBase> const&, NYT::TIntrusivePtr<NYT::NControllerAgent::TSortOperationOptionsBase> const&, NYT::NLogging::TLogger&, long&, long&, long&, double&>(void*, NYT::TIntrusivePtr<NYT::NScheduler::TSortOperationSpecBase> const&, NYT::TIntrusivePtr<NYT::NControllerAgent::TSortOperationOptionsBase> const&, NYT::NLogging::TLogger&, long&, long&, long&, double&) at /home/teamcity/source/Yt_ArcRelease/library/cpp/yt/memory/new-inl.h:158
 6. 0x0000156a997a in NYT::TIntrusivePtr<NYT::NControllerAgent::TPartitionJobSizeConstraints> NYT::NDetail::SafeConstruct<NYT::NControllerAgent::TPartitionJobSizeConstraints, NYT::TIntrusivePtr<NYT::NScheduler::TSortOperationSpecBase> const&, NYT::TIntrusivePtr<NYT::NControllerAgent::TSortOperationOptionsBase> const&, NYT::NLogging::TLogger&, long&, long&, long&, double&>(void*, NYT::TIntrusivePtr<NYT::NScheduler::TSortOperationSpecBase> const&, NYT::TIntrusivePtr<NYT::NControllerAgent::TSortOperationOptionsBase> const&, NYT::NLogging::TLogger&, long&, long&, long&, double&) at /home/teamcity/source/Yt_ArcRelease/library/cpp/yt/memory/new-inl.h:168
 7. 0x0000156a997a in NYT::TIntrusivePtr<NYT::NControllerAgent::TPartitionJobSizeConstraints> NYT::New<NYT::NControllerAgent::TPartitionJobSizeConstraints, NYT::TIntrusivePtr<NYT::NScheduler::TSortOperationSpecBase> const&, NYT::TIntrusivePtr<NYT::NControllerAgent::TSortOperationOptionsBase> const&, NYT::NLogging::TLogger&, long&, long&, long&, double&, void>(NYT::TIntrusivePtr<NYT::NScheduler::TSortOperationSpecBase> const&, NYT::TIntrusivePtr<NYT::NControllerAgent::TSortOperationOptionsBase> const&, NYT::NLogging::TLogger&, long&, long&, long&, double&) at /home/teamcity/source/Yt_ArcRelease/library/cpp/yt/memory/new-inl.h:203
 8. 0x0000156a997a in NYT::NControllerAgent::CreatePartitionJobSizeConstraints(NYT::TIntrusivePtr<NYT::NScheduler::TSortOperationSpecBase> const&, NYT::TIntrusivePtr<NYT::NControllerAgent::TSortOperationOptionsBase> const&, NYT::NLogging::TLogger, long, long, long, double) at /home/teamcity/source/Yt_ArcRelease/yt/yt/server/controller_agent/job_size_constraints.cpp:816
 9. 0x00001578bd92 in NYT::NControllerAgent::NControllers::TMapReduceController::CustomMaterialize() at /home/teamcity/source/Yt_ArcRelease/yt/yt/server/controller_agent/controllers/sort_controller.cpp:4257
10. 0x0000155507c7 in NYT::NControllerAgent::NControllers::TOperationControllerBase::SafeMaterialize() at /home/teamcity/source/Yt_ArcRelease/yt/yt/server/controller_agent/controllers/operation_controller_detail.cpp:1231
11. 0x0000154ed5df in NYT::NControllerAgent::NControllers::TOperationControllerBase::Materialize() at /home/teamcity/source/Yt_ArcRelease/yt/yt/server/controller_agent/controllers/operation_controller_detail.h:159
12. 0x0000154de222 in decltype(auto) NYT::NControllerAgent::TOperationControllerWrapper::DoExecuteGuarded<NYT::NControllerAgent::IOperationControllerSchedulerHost, NYT::NControllerAgent::TOperationControllerMaterializeResult>(NYT::NControllerAgent::TOperationControllerMaterializeResult (NYT::NControllerAgent::IOperationControllerSchedulerHost::*)()) at /home/teamcity/source/Yt_ArcRelease/yt/yt/server/controller_agent/operation_controller.cpp:164
13. 0x0000154daa14 in NYT::NControllerAgent::TOperationControllerWrapper::Materialize() at /home/teamcity/source/Yt_ArcRelease/yt/yt/server/controller_agent/operation_controller.cpp:257
14. 0x0000151e4067 in auto NYT::NDetail::TMethodInvoker<NYT::NControllerAgent::TOperationControllerMaterializeResult (NYT::NControllerAgent::IOperationControllerSchedulerHost::*)()>::operator()<NYT::NControllerAgent::IOperationController>(NYT::TIntrusivePtr<NYT::NControllerAgent::IOperationController> const&) const at /home/teamcity/source/Yt_ArcRelease/yt/yt/core/actions/bind-inl.h:257
15. 0x0000151e4067 in auto NYT::NDetail::TBindState<true, NYT::NDetail::TMethodInvoker<NYT::NControllerAgent::TOperationControllerMaterializeResult (NYT::NControllerAgent::IOperationControllerSchedulerHost::*)()>, std::__y1::integer_sequence<unsigned long, 0ul>, NYT::TIntrusivePtr<NYT::NControllerAgent::IOperationController> >::Run<>(NYT::NDetail::TBindStateBase*) at /home/teamcity/source/Yt_ArcRelease/yt/yt/core/actions/bind-inl.h:510
16. 0x0000151e60a6 in NYT::TCallback<NYT::NControllerAgent::TOperationControllerMaterializeResult ()>::operator()() const at /home/teamcity/source/Yt_ArcRelease/yt/yt/core/actions/callback.h:234
17. 0x0000151e60a6 in void NYT::NDetail::TPromiseSetter<NYT::NControllerAgent::TOperationControllerMaterializeResult, NYT::NControllerAgent::TOperationControllerMaterializeResult ()>::Do<>(NYT::TPromise<NYT::NControllerAgent::TOperationControllerMaterializeResult> const&, NYT::TCallback<NYT::NControllerAgent::TOperationControllerMaterializeResult ()> const&)::'lambda'()::operator()() const at /home/teamcity/source/Yt_ArcRelease/yt/yt/core/actions/future-inl.h:717
18. 0x0000151e60a6 in void NYT::NDetail::InterceptExceptions<NYT::NControllerAgent::TOperationControllerMaterializeResult, void NYT::NDetail::TPromiseSetter<NYT::NControllerAgent::TOperationControllerMaterializeResult, NYT::NControllerAgent::TOperationControllerMaterializeResult ()>::Do<>(NYT::TPromise<NYT::NControllerAgent::TOperationControllerMaterializeResult> const&, NYT::TCallback<NYT::NControllerAgent::TOperationControllerMaterializeResult ()> const&)::'lambda'()>(NYT::TPromise<NYT::NControllerAgent::TOperationControllerMaterializeResult> const&, void NYT::NDetail::TPromiseSetter<NYT::NControllerAgent::TOperationControllerMaterializeResult, NYT::NControllerAgent::TOperationControllerMaterializeResult ()>::Do<>(NYT::TPromise<NYT::NControllerAgent::TOperationControllerMaterializeResult> const&, NYT::TCallback<NYT::NControllerAgent::TOperationControllerMaterializeResult ()> const&)::'lambda'() const&) at /home/teamcity/source/Yt_ArcRelease/yt/yt/core/actions/future-inl.h:698
19. 0x0000151e47d8 in void NYT::NDetail::TPromiseSetter<NYT::NControllerAgent::TOperationControllerMaterializeResult, NYT::NControllerAgent::TOperationControllerMaterializeResult ()>::Do<>(NYT::TPromise<NYT::NControllerAgent::TOperationControllerMaterializeResult> const&, NYT::TCallback<NYT::NControllerAgent::TOperationControllerMaterializeResult ()> const&) at /home/teamcity/source/Yt_ArcRelease/yt/yt/core/actions/future-inl.h:714
20. 0x0000151e47d8 in NYT::NDetail::TAsyncViaHelper<NYT::NControllerAgent::TOperationControllerMaterializeResult ()>::Inner(NYT::TExtendedCallback<NYT::NControllerAgent::TOperationControllerMaterializeResult ()> const&, NYT::TPromise<NYT::NControllerAgent::TOperationControllerMaterializeResult> const&) at /home/teamcity/source/Yt_ArcRelease/yt/yt/core/actions/future-inl.h:1576
21. 0x00000fbdc932 in NYT::TCallback<void ()>::operator()() const at /home/teamcity/source/Yt_ArcRelease/yt/yt/core/actions/callback.h:234
22. 0x00000fbdc932 in NYT::TCancelableContext::TCancelableInvoker::Invoke(NYT::TCallback<void ()>)::'lambda'()::operator()() const at /home/teamcity/source/Yt_ArcRelease/yt/yt/core/actions/cancelable_context.cpp:33
23. 0x00000fbdc932 in auto NYT::NDetail::TBindState<false, NYT::TCancelableContext::TCancelableInvoker::Invoke(NYT::TCallback<void ()>)::'lambda'(), std::__y1::integer_sequence<unsigned long> >::Run<>(NYT::NDetail::TBindStateBase*) at /home/teamcity/source/Yt_ArcRelease/yt/yt/core/actions/bind-inl.h:510
*** Marked memory regions of total size 0 MB as undumpable ***
*** Waiting for logger to shut down ***
*** Terminating ***`

**Comment by gritukan** [comment]
This is not surprising since old controller agent with a bug can be used in compat test. I've added a check that trunk controller agent is used in tests.

**Comment by psushin** [comment]
Shipped!

**Pull request closed without merging**